src/thor/dijkstras.cc

```diff
@@ -154,34 +159,71 @@ void Dijkstras::ExpandForward(GraphReader& graphreader,
     // (based on the costing method) or if a complex restriction exists for
     // this path.
     if (directededge->is_shortcut() || es->set() == EdgeSet::kPermanent ||
-        !(directededge->forwardaccess() & access_mode_)) {
+        !((FORWARD ? directededge->forwardaccess() : directededge->reverseaccess()) & access_mode_)) {
       continue;
     }
 
+    // Get end node tile, opposing edge Id, and opposing directed edge.
+    graph_tile_ptr t2;
+    baldr::GraphId oppedgeid;
+    const baldr::DirectedEdge* opp_edge;
+    // Only populate these at this point for the reverse search - the forward search
+    // may exit before needing them, so we can defer some work until later
+    if (not FORWARD) { // aka reverse
+      t2 = tile;
+      oppedgeid = graphreader.GetOpposingEdgeId(edgeid, t2);
+      if (t2 == nullptr) {
+        continue;
+      }
+      opp_edge = t2->directededge(oppedgeid);
+    }
+
     // Check if the edge is allowed or if a restriction occurs
     EdgeStatus* todo = nullptr;
     uint8_t restriction_idx = -1;
     if (offset_time.valid) {
       // With date time we check time dependent restrictions and access
-      if (!costing_->Allowed(directededge, pred, tile, edgeid, offset_time.local_time,
-                             nodeinfo->timezone(), restriction_idx) ||
-          costing_->Restricted(directededge, pred, bdedgelabels_, tile, edgeid, true, todo,
-                               offset_time.local_time, nodeinfo->timezone())) {
+      const bool allowed =
+          FORWARD ? costing_->Allowed(directededge, pred, tile, edgeid, offset_time.local_time,
+                                      nodeinfo->timezone(), restriction_idx)
+                  : costing_->AllowedReverse(directededge, pred, opp_edge, t2, oppedgeid,
+                                             offset_time.local_time, nodeinfo->timezone(),
+                                             restriction_idx);
```

Comment:
this almost makes me want to see the templatization trickle down into costing in the future. one step at a time though haha

Reply:
Yeah, I had exactly the same though.  I've bookmarked it for later in the refactor :-)